src/daft-local-execution/src/intermediate_ops/udf.rs

```diff
@@ -0,0 +1,366 @@
+        Python::with_gil(|py| {
+            Ok(inner
+                .bind(py)
+                .call_method1(
+                    pyo3::intern!(py, "eval_input"),
+                    (PyMicroPartition::from(input),),
+                )?
+                .extract::<PyMicroPartition>()?
+                .into())
+        })
+    }
+
+    #[cfg(feature = "python")]
+    fn eval_input_inline(&mut self, input: Arc<MicroPartition>) -> DaftResult<Arc<MicroPartition>> {
+        // Extract the udf_expr into a PythonUDF and optional name
+        let inner_expr = self.udf_expr.inner();
+        let (inner_expr, out_name) = inner_expr.unwrap_alias();
+        let Expr::Function {
+            func: FunctionExpr::Python(func),
+            inputs: input_exprs,
+        } = inner_expr.as_ref()
+        else {
+            return Err(DaftError::InternalError(format!(
+                "Expected a Python UDF, got {}",
+                inner_expr
+            )));
+        };
+
+        let input_exprs = BoundExpr::bind_all(input_exprs, input.schema().as_ref())?;
+
+        let mut out_batches = vec![];
+
+        // Iterate over MicroPartition batches
+        for batch in input.get_tables()?.as_ref() {
+            // Get the functions inputs
+            let func_input = batch.eval_expression_list(input_exprs.as_slice())?;
+            // Call the UDF, getting the GIL contention time
+            let (mut result, gil_contention_time) = func.call_udf(func_input.columns())?;
+            if let Some(out_name) = out_name.as_ref() {
+                result = result.rename(out_name);
+            }
+
+            // Update the state
+            self.total_runtime += gil_contention_time;
+            self.total_gil_contention += gil_contention_time;
+            self.num_batches += 1;
+
+            let passthrough_input =
+                batch.eval_expression_list(self.passthrough_columns.as_slice())?;
+            let series = passthrough_input.append_column(result)?;
+            out_batches.push(series);
+        }
+
+        // Switch to external process if we hit the GIL contention threshold
+        if self.num_batches > 10
+            && (self.total_gil_contention.as_secs_f64() / self.total_runtime.as_secs_f64()) > 0.2
+        {
+            self.create_handle()?;
+        }
+
+        let out_schema = out_batches[0].schema.clone();
+        Ok(Arc::new(MicroPartition::new_loaded(
+            out_schema,
+            Arc::new(out_batches),
+            None,
+        )))
+    }
+
+    fn eval_input(&mut self, input: Arc<MicroPartition>) -> DaftResult<Arc<MicroPartition>> {
+        #[cfg(not(feature = "python"))]
+        {
+            panic!("Cannot evaluate a UDF without compiling for Python");
+        }
+
+        #[cfg(feature = "python")]
+        {
+            if let Some(inner) = &self.inner {
+                self.eval_input_with_handle(input, inner)
+            } else {
+                self.eval_input_inline(input)
+            }
```

Comment:
This is the crux of the runner / optimization. The ability to start running a UDF inline in the same thread but later switch to another process if GIL overhead is too high. See the helper functions for more details